src/common/libevent.hpp

```diff
@@ -463,6 +466,150 @@ class Bufferevent : public NonCopyable, public NonMovable {
     }
 };
 
+class EvdnsBase {
+  public:
+    // Add a reference to Var<EventBase> base to make sure it is referenced
+    Var<EventBase> evbase = nullptr;
+    evdns_base *dns_base = nullptr;
+
+    EvdnsBase() {}
+    ~EvdnsBase() {}
+
+    template <decltype(evdns_base_new) construct = ::evdns_base_new,
+              decltype(evdns_base_free) destruct = ::evdns_base_free>
+    static Var<EvdnsBase> create(Var<EventBase> base,
+                                 bool initialize_nameservers = true,
+                                 bool fail_requests = true) {
+
+        auto pointer = construct(base->evbase, initialize_nameservers);
+        if (pointer == nullptr) {
+            MK_THROW(EvdnsBaseNewExceptionError);
+        }
+        Var<EvdnsBase> _base(new EvdnsBase, [fail_requests](EvdnsBase *ptr) {
+            destruct(ptr->dns_base, fail_requests);
+            delete ptr;
+        });
+        _base->evbase = base;
+        _base->dns_base = pointer;
+        return _base;
+    }
+
+    typedef std::function<void(int result, char type, int count, int ttl,
+                               std::vector<std::string> addresses)>
+        ResolveCallback;
+
+    static std::vector<std::string> ipv4_address_list(int count,
+                                                      void *addresses) {
+        std::vector<std::string> results;
+        static const int size = 4;
+        if (count >= 0 && count <= INT_MAX / size + 1) {
+            char string[128]; // Is wide enough (max. IPv6 length is 45 chars)
+            for (int i = 0; i < count; ++i) {
+                // Note: address already in network byte order
+                if (inet_ntop(AF_INET, (char *)addresses + i * size, string,
+                              sizeof(string)) == nullptr) {
+                    break;
+                }
+                results.push_back(string);
+            }
+        }
+        return results;
+    }
+
+    typedef std::function<void(int, char, int, int, void *)> evdns_callback;
+    // TODO: C callbacks should be declared with C linkage
```

Comment:
This basically means to add the prototype at the top of this file and the implementation in `src/common/libevent.cpp`. Also, since C linkage implies no name mangling, I'd call this function `mk_evdns_handle_resolve()`.